src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs

```diff
@@ -817,6 +1025,32 @@ public struct S
                 Diagnostic(ErrorCode.ERR_BadMemberFlag, "S").WithArguments("readonly").WithLocation(5, 21));
         }
 
+        [Fact]
+        public void ReadOnlyStruct_Constructor()
+        {
+            var csharp = @"
+public readonly struct S
+{
+    public readonly int i;
+    public S(int i)
+    {
+        this.i = i; // ok
+        M(ref this); // ok
+    }
+    public static void M(ref S s)
+    {
+        s.i = 42; // error
+        s = default; // ok
+    }
+}
+";
+            var comp = CreateCompilation(csharp);
+            comp.VerifyDiagnostics(
+                // (12,9): error CS0191: A readonly field cannot be assigned to (except in a constructor or a variable initializer)
+                //         s.i = 42; // error
+                Diagnostic(ErrorCode.ERR_AssgReadonly, "s.i").WithLocation(12, 9));
+        }
```

Comment:
>} [](start = 8, length = 1)

Please add tests for methods that are called implicitly:
- user-defined operators
- `foreach` and `GetEnumerator()`
- `using` and `Dispose()`

Reply:
- Agree there's not much more to test with user-defined operators because they're static methods.
- Will test that `foreach (var x in this)` warns appropriately within a readonly method
- Will test that `using (this) {}` warns appropriately within a readonly method
- Will test that deconstruction warns appropriately within a readonly method

Reply:
May need to also test `await foreach` and `DisposeAsync`

Reply:
Done